stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -0,0 +1,71 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tracing implements the OpenTelemetry carrier for context propagation
+// in gRPC tracing.
+package tracing
+
+import (
+	"google.golang.org/grpc/metadata"
+)
+
+// CustomCarrier is a TextMapCarrier that uses `*metadata.MD` to store and
+// retrieve any propagated key-value pairs in text format.
+type CustomCarrier struct {
+	md *metadata.MD
+}
+
+// NewCustomCarrier creates a new CustomCarrier with
+// the given metadata.
+func NewCustomCarrier(md *metadata.MD) *CustomCarrier {
```

Comment:
metadata.MD is a map.  Maps are pointers already, so a pointer to a map is a double dereference.

Reply:
Hmm, thinking about the other PR.  Putting a map in here means we can't use the more optimized `AppendToOutgoingContext`.  Since this is all RPC path stuff, I think it might be better to go back to contexts and use `Append` and `ValueFrom` instead.  Otherwise we suffer a copy of the whole metadata map when adding to / pulling from the context (in the metadata package).

Reply:
yeah that's fair. Changed back to context approach and used efficient AppendToOutgoingContext and ValueFromIncomingContext